### ISSUE TITLE
Access all Attributes of records or columns of table

### ISSUE THREAD

**Issue by combinatorist** [gristlabs/grist-core#629]
I would like a way to save the entire record as it exited at a particular moment as a python object.

Barrier # 1:
- Trigger functions can't use the `Any` type.
- workaround: Convert to a json string instead of a python object

Barrier # 2:
- The `rec` only contains a shallow reference to its location (table and row number), so no data to json-stringify.
- workaround: Fine, I'll just access the column names via the table object and programmatically call each one ...

Barrier # 3:
- The table object somehow does note have the `all_columns` method available (maybe that's a different table class?).

And now I'm stuck. How can I programmatically dump out all the data in a record at a given moment (via trigger function)?

**Comment by paulfitz** [comment]
Is the [RECORD](https://support.getgrist.com/functions/#record_2) function of any use?

**Comment by alexmojaki** [comment]
> The table object somehow does note have the all_columns method available (maybe that's a different table class?)

The tables 'directly' accessible to formulas are `UserTable`s, the underlying `Table` is accessible via `.table`, so e.g. `Table1.table.all_columns` should work, although it will probably contain stuff you don't want.

**Comment by combinatorist** [comment]
> Is the [RECORD](https://support.getgrist.com/functions/#record_2) function of any use?

Lol, should have known there would be a function for that. I originally tried to make `rec` work and didn't look for anything else after that. As for the naming convention, it's a little weird that it looks like an XL function with the all caps right? Why wouldn't it be `record`, `asDict`, or even `rec.__dict__`? Just food for thought.

**Comment by paulfitz** [comment]
Agree it is named weirdly and is hard to discover; I for one would look kindly on a tasteful PR with a good pythony entrypoint for the functionality. It came into existence as a workaround for some awkwardness in implementing a custom widget for invoicing.

**Comment by combinatorist** [comment]
> Is the [RECORD](https://support.getgrist.com/functions/#record_2) function of any use?

Actually, I've just found that it's not really working super well for me. My more general problem is that I'd like to keep a history of versions of each record. At each trigger function, I would just append each result to a python list, but then the trigger function turns it (sort of) into a string, so it won't let me append as a list. So, my next move was to Ser/De with json, but it gave an hard to trace error about RECORD not being JSON serializable.

I finally realized since it has to be a string anyway, if I postpone parsing and even proper serialization for now, at least I can use python's `str`, so I'm just appending a newline and a representation of the record each time a column change triggers the function. It's mostly a bunch of ref columns with meaningless indices, but if I take care not to change record ids(!), at least a human will be able to track the changes if absolutely necessary later. I could also make trigger functions to spit out better ids for each reference column, but I'm not bothering now. :) :( :/

Thanks for your help, @paulfitz!
